ext/drreg/drreg.c

```diff
@@ -68,14 +69,27 @@
 /* This should be pretty hard to exceed as there aren't this many GPRs */
 #define MAX_SPILLS (SPILL_SLOT_MAX + 8)
 
-#define AFLAGS_SLOT 0 /* always */
+/* This should be pretty hard to exceed as there aren't this many SIMD registers */
+#define MAX_SIMD_SPILLS (MCXT_NUM_SIMD_SLOTS + 8)
+#define REG_SIMD_SIZE 64
+#define XMM_REG_SIZE 16
 
-/* We support using GPR registers only: [DR_REG_START_GPR..DR_REG_STOP_GPR] */
+#define AFLAGS_SLOT 0 /* always */
 
+/* Liveliness states for gprs */
 #define REG_DEAD ((void *)(ptr_uint_t)0)
 #define REG_LIVE ((void *)(ptr_uint_t)1)
 #define REG_UNKNOWN ((void *)(ptr_uint_t)2) /* only used outside drmgr insert phase */
 
+/* Liveliness states for SIMD (not for mmx) */
+#define SIMD_XMM_DEAD ((void *)(ptr_uint_t)0) /* first 16 bytes are dead, rest a live */
```

Comment:
I am not sure I quite understand what is the problem. The idea here is to have specific liveliness states so that drreg can then intelligently spill and restore partial SIMD registers. If we want to reserve just an XMM register and we note that the register has a liveliness state of SIMD_XMM_DEAD, then we don't need to spill. In contrast, if we want to reserve a YMM register then drreg needs to do spillage. These are similar to the GPR states, namely REG_DEAD and REG_LIVE. The only difference is that they are for SIMD vector registers.

Reply:
If you are just suggesting an alternative approach, where we use a bit-vector, then we need to see how we can handle partial register storage, particularly for SIMD registers. The bit-vector would require storing a flag for each register byte in this case? Alternatively, we make the decision to always spill the largest SIMD vector (ZMM) register despite the user only needing an XMM register - we need to handle non-AVX-512 architectures.

Note, drreg currently does not handle partial spillage for GPRs such as saving AX (16-bit). However, for GPRs, I don't think we should, as we can face perf issues, e.g., due to partial register stalls.

Reply:
Not sure why there are 3 states, live, dead, and unknown. 'unknown' seems to be used in forward liveness analysis. I would think 1 bit, either dead or live (conservative) would be enough. Can you have a look and determine why it is needed, will look later, too.

There is another problem. Only non-VEX writes to xmm do not null zmmh. All others make zmm 'dead'. Same for all writes to ymm, they are all vex or evex, unless we want to examine a potential avx-512 mask as well (we don't). But before we spill/restore zmms when vex instruction are encountered, we would need to test for proc_avx512_enabled(). Then OTOH, we probably don't want to introduce avx-512 spill/fill to zmm, if the app is only SSE+AVX. Context switching handles this by inserting a big if/else based on d_r_avx512_code_in_use, but we prob. don't want or can't insert this here.

If you like to limit this PR to being able to handle xmm only, that would totally be fine. There is actually a use case for being able to spill opmask registers (#2985), which is more important. If you do want to include ymm and zmm, need to think about above.

Reply:
If just xmms are considered and spill/restore is using SSE movdqa, I think there is no problem. Problem begins when liveness of ymm or zmm needs to be tracked. Several issues: 

1) Figuring out which writes to xmm null ymmh, zmmh. SSE don't, all VEX and EVEX do.
2) Might even depend on AVX-512 mask whether they _not_ write to ymmh, zmmh, though you could conservatively assume that all non-VEX always do.
3) Scenario: a client wants to reserve ymm. Just spill/restore ymm with avx code nulls zmmh. But we don't always want to insert AVX-512 spill/restore in this case, because we don't want to trash a SSE or AVX application with AVX-512 code for performance reasons. A great effort was spent to lazily context switch for example (search for recent git history). In order to know whether zmms may be live, context switching looks at _d_r_avx512_code_in_use_ which is lazily set upon AVX-512 code detection. In order to be correct, the check needs to be dynamic, i.e. an if-statement needs to be in the code cache. I don't think this is a good idea here for drreg.

So maybe slns could be

1) Check for non-VEX, instr_is_sse().
2) Be conservative.
3) Refer to proc_avx512_enabled() and tell the user that we _will_ have to insert AVX-512 zmm spill/restore for ymm, even if the app is not AVX-512, but processor + OS are, and comment accordingly.

Thinking of other cases..

Reply:
> Is it mainly to use a bit-vector which leads to better code clarity?
No, that's separate (still not sure why we need 3 states, even in existing code)

Reply:
Okay I understand. I still need to create a test with the scenario you have mentioned. However, assuming we are fine with thrashing AVX-512 instructions regardless whether or not the app uses such instructions for its own operations, I think the main challenge of addressing this issue is close to being handled in the current approach. Essentially, if an app uses VEX/EVEX instructions and the register is LIVE, we spill the entire ZMM register. Assuming instr_writes_to_reg differentiates between SSE and VEX/EVEX writes to vector registers, I think we should be okay no?

Reply:
> However, assuming we are fine with thrashing AVX-512 instructions regardless whether or not the app uses such instructions for its own operations

We are not, please re-read. You're probably referring to 3. [above](https://github.com/DynamoRIO/dynamorio/pull/3893/#discussion_r335687607), right? What this means is, in order to _not_ trash live application registers (never), we would non-lazily always insert zmm spill/fill for spilling ymm, as long as proc_avx512_enabled(), and comment that in drreg's description. Because lazily doing this would require inserting complicated control flow that will have its own issues. We could also add a flag for the user to assert that the application is AVX only and does not use AVX-512.

Reply:
Yes, I think we are on the same page - I used the work trash to refer to AVX-512 code not registers in reference to this:

> 3\. trash a SSE or AVX application with AVX-512 code for performance reasons

Reply:
Ok.

Reply:
W.r.t. bitvector and why REG_UNKNOWN (I guess I am talking to myself since I asked the question in the first place): when we're doing a forward liveness scan when reserve is called outside of drmgr's insertion phase, it's a way to speed up analysis. Instead of scanning the entire basic block for whether the register is dead, we can stop once we found it is live. So you can live this methodology in the code and do the same for [xyz]mm, it seems fine.